File: mod/ondb/lib/py/compact.py
"""Compact TSV log: output current state as snapshot TSV."""
import sys, os
sys.path.insert(0, os.path.dirname(__file__))
from ondb_core import Ondb

db = Ondb()
db.load(sys.stdin)

# Entities: add + optional set for mtime
for eid in db._entity_order:
    e = db.entities[eid]
    if not e['alive']:
        continue
    cols = ['add', e['type'], eid, str(e['ctime'])]
    for k in e['props'].keys():
        v = e['props'][k]
        # Escape tabs/newlines in values like AWK _onto_escape
        v = v.replace('\\', '\\\\').replace('\t', '\\t').replace('\n', '\\n')
        cols.append(k)
        cols.append(v)
    print('\t'.join(cols))
    if e['mtime'] != e['ctime']:
        print(f"set\t{eid}\t{e['mtime']}")

# Links
for link in db.links:
    if link['from'] is None:
        continue
    cols = ['link', link['from'], link['rel'], link['to'], str(link['epoch'])]
    for k in link['props'].keys():
        v = link['props'][k]
        v = v.replace('\\', '\\\\').replace('\t', '\\t').replace('\n', '\\n')
        cols.append(k)
        cols.append(v)
    print('\t'.join(cols))

File: mod/ondb/lib/js/compact.js
import { Ondb } from './ondb_core.js';

const db = new Ondb();
const input = await Bun.stdin.text();
db.load(input);

const out = [];

// Entities
for (let i = 0; i < db._entity_order.length; i++) {
  const eid = db._entity_order[i];
  const e = db.entities.get(eid);
  if (!e || !e.alive) continue;
  const parts = ['add', e.type, eid, String(e.ctime)];
  for (const k in e.props) {
    let v = e.props[k];
    if (v.indexOf('\\') >= 0 || v.indexOf('\t') >= 0 || v.indexOf('\n') >= 0) {
      v = v.replaceAll('\\', '\\\\').replaceAll('\t', '\\t').replaceAll('\n', '\\n');
    }
    parts.push(k, v);
  }
  out.push(parts.join('\t'));
  if (e.mtime !== e.ctime) {
    out.push(`set\t${eid}\t${e.mtime}`);
  }
}

// Links
for (let i = 0; i < db.links.length; i++) {
  const link = db.links[i];
  if (link.from === null) continue;
  const parts = ['link', link.from, link.rel, link.to, String(link.epoch)];
  for (const k in link.props) {
    let v = link.props[k];
    if (v.indexOf('\\') >= 0 || v.indexOf('\t') >= 0 || v.indexOf('\n') >= 0) {
      v = v.replaceAll('\\', '\\\\').replaceAll('\t', '\\t').replaceAll('\n', '\\n');
    }
    parts.push(k, v);
  }
  out.push(parts.join('\t'));
}

process.stdout.write(out.join('\n') + '\n');
